File: overlay_window.py
"""
Fullscreen Overlay Window for displaying messages or blocking screen
"""
import tkinter as tk
from tkinter import font


class OverlayWindow:
    """Fullscreen overlay window that can display messages"""

    def __init__(self):
        self.window = None
        self.is_showing = False

    def show(self, message="", bg_color="black", text_color="white", opacity=0.9):
        """
        Show fullscreen overlay window

        Args:
            message: Text message to display (optional)
            bg_color: Background color (default: black)
            text_color: Text color (default: white)
            opacity: Window opacity 0.0-1.0 (default: 0.9)
        """
        if self.is_showing:
            self.hide()

        self.window = tk.Tk()
        self.window.title("Remote Control Overlay")

        # Configure fullscreen
        self.window.attributes('-fullscreen', True)
        self.window.attributes('-topmost', True)
        self.window.attributes('-alpha', opacity)

        # Set background color
        self.window.configure(bg=bg_color)

        # Create main frame
        frame = tk.Frame(self.window, bg=bg_color)
        frame.pack(expand=True, fill='both')

        if message:
            # Display message
            msg_font = font.Font(family='Arial', size=48, weight='bold')
            label = tk.Label(
                frame,
                text=message,
                font=msg_font,
                fg=text_color,
                bg=bg_color,
                wraplength=1200
            )
            label.pack(expand=True)

        # Add close instruction
        close_font = font.Font(family='Arial', size=16)
        close_label = tk.Label(
            frame,
            text="Press ESC to close",
            font=close_font,
            fg=text_color,
            bg=bg_color
        )
        close_label.pack(side='bottom', pady=20)

        # Bind ESC key to close
        self.window.bind('<Escape>', lambda e: self.hide())

        # Bind click to close
        self.window.bind('<Button-1>', lambda e: self.hide())

        self.is_showing = True

        # Start event loop
        self.window.mainloop()

    def show_async(self, message="", bg_color="black", text_color="white", opacity=0.9):
        """Show overlay in a separate thread"""
        import threading
        thread = threading.Thread(
            target=self.show,
            args=(message, bg_color, text_color, opacity),
            daemon=True
        )
        thread.start()

    def hide(self):
        """Hide the overlay window"""
        if self.window and self.is_showing:
            try:
                self.window.quit()
                self.window.destroy()
            except:
                pass
            finally:
                self.window = None
                self.is_showing = False

    def show_blocking_screen(self, message="Screen Locked"):
        """Show a blocking screen overlay (dark overlay)"""
        self.show_async(message=message, bg_color="black", text_color="red", opacity=0.95)

    def show_notification(self, message):
        """Show a notification overlay (semi-transparent)"""
        self.show_async(message=message, bg_color="blue", text_color="white", opacity=0.7)

    def show_warning(self, message):
        """Show a warning overlay"""
        self.show_async(message=message, bg_color="orange", text_color="black", opacity=0.85)

    def show_error(self, message):
        """Show an error overlay"""
        self.show_async(message=message, bg_color="red", text_color="white", opacity=0.9)

File: WindowsRemoteTools/OverlayWindow.cs
using System;
using System.Drawing;
using System.IO;
using System.Threading;
using System.Windows.Forms;

namespace WindowsRemoteTools
{
    /// <summary>
    /// Fullscreen overlay window for displaying messages and pictures
    /// </summary>
    public class OverlayWindow
    {
        private Form? _currentForm;
        private readonly object _lock = new object();

        public void Show(string message = "", Color? backgroundColor = null, Color? textColor = null, double opacity = 0.9)
        {
            // Must run on UI thread
            if (_currentForm?.InvokeRequired ?? false)
            {
                _currentForm.Invoke(new Action(() => Show(message, backgroundColor, textColor, opacity)));
                return;
            }

            Hide();

            var bgColor = backgroundColor ?? Color.Black;
            var txtColor = textColor ?? Color.White;

            var form = new Form
            {
                FormBorderStyle = FormBorderStyle.None,
                WindowState = FormWindowState.Maximized,
                TopMost = true,
                BackColor = bgColor,
                Opacity = opacity,
                ShowInTaskbar = false,
                StartPosition = FormStartPosition.Manual,
                Bounds = Screen.PrimaryScreen?.Bounds ?? Screen.AllScreens[0].Bounds
            };

            // Main panel
            var panel = new Panel
            {
                Dock = DockStyle.Fill,
                BackColor = bgColor
            };

            if (!string.IsNullOrEmpty(message))
            {
                var label = new Label
                {
                    Text = message,
                    Font = new Font("Arial", 48, FontStyle.Bold),
                    ForeColor = txtColor,
                    BackColor = bgColor,
                    AutoSize = false,
                    TextAlign = ContentAlignment.MiddleCenter,
                    Dock = DockStyle.Fill
                };
                panel.Controls.Add(label);
            }

            // Close instruction
            var closeLabel = new Label
            {
                Text = "Press ESC or click to close",
                Font = new Font("Arial", 16),
                ForeColor = txtColor,
                BackColor = bgColor,
                AutoSize = true,
                TextAlign = ContentAlignment.BottomCenter,
                Dock = DockStyle.Bottom,
                Height = 50,
                Padding = new Padding(0, 0, 0, 20)
            };
            panel.Controls.Add(closeLabel);

            form.Controls.Add(panel);

            // Event handlers
            form.KeyDown += (s, e) =>
            {
                if (e.KeyCode == Keys.Escape)
                    Hide();
            };

            form.Click += (s, e) => Hide();
            panel.Click += (s, e) => Hide();

            lock (_lock)
            {
                _currentForm = form;
            }

            form.Show();
        }

        public void ShowAsync(string message = "", Color? backgroundColor = null, Color? textColor = null, double opacity = 0.9)
        {
            var thread = new Thread(() =>
            {
                Application.EnableVisualStyles();
                Show(message, backgroundColor, textColor, opacity);
                Application.Run();
            });
            thread.SetApartmentState(ApartmentState.STA);
            thread.IsBackground = true;
            thread.Start();
        }

        public void Hide()
        {
            lock (_lock)
            {
                if (_currentForm != null)
                {
                    try
                    {
                        if (_currentForm.InvokeRequired)
                        {
                            _currentForm.Invoke(new Action(() =>
                            {
                                _currentForm.Close();
                                _currentForm.Dispose();
                            }));
                        }
                        else
                        {
                            _currentForm.Close();
                            _currentForm.Dispose();
                        }
                    }
                    catch { }
                    finally
                    {
                        _currentForm = null;
                    }
                }
            }
        }

        public void ShowBlockingScreen(string message = "Screen Locked")
        {
            ShowAsync(message, Color.Black, Color.Red, 0.95);
        }

        public void ShowLockedScreen(string message, string passwordHash, Action? onUnlocked = null)
        {
            Hide();
            var thread = new Thread(() =>
            {
                Application.EnableVisualStyles();
                var form = new LockedOverlayForm(message, passwordHash, onUnlocked);
                Application.Run(form);
            });
            thread.SetApartmentState(ApartmentState.STA);
            thread.IsBackground = true;
            thread.Start();
        }

        public void ShowNotification(string message)
        {
            ShowAsync(message, Color.Blue, Color.White, 0.7);
        }

        public void ShowWarning(string message)
        {
            ShowAsync(message, Color.Orange, Color.Black, 0.85);
        }

        public void ShowError(string message)
        {
            ShowAsync(message, Color.Red, Color.White, 0.9);
        }

        /// <summary>
        /// Shows a picture in fullscreen overlay
        /// </summary>
        public void ShowPicture(string imagePath, bool maintainAspectRatio = true, Color? backgroundColor = null)
        {
            var thread = new Thread(() =>
            {
                try
                {
                    Application.EnableVisualStyles();

                    if (!File.Exists(imagePath))
                    {
                        Console.WriteLine($"Picture not found: {imagePath}");
                        ShowError("Picture not found");
                        return;
                    }

                    Hide();

                    var bgColor = backgroundColor ?? Color.Black;

                    var form = new Form
                    {
                        FormBorderStyle = FormBorderStyle.None,
                        WindowState = FormWindowState.Maximized,
                        TopMost = true,
                        BackColor = bgColor,
                        ShowInTaskbar = false,
                        StartPosition = FormStartPosition.Manual,
                        Bounds = Screen.PrimaryScreen?.Bounds ?? Screen.AllScreens[0].Bounds
                    };

                    // Create PictureBox for image display
                    var pictureBox = new PictureBox
                    {
                        Dock = DockStyle.Fill,
                        BackColor = bgColor,
                        SizeMode = maintainAspectRatio ? PictureBoxSizeMode.Zoom : PictureBoxSizeMode.StretchImage
                    };

                    try
                    {
                        pictureBox.Image = Image.FromFile(imagePath);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"Error loading image: {ex.Message}");
                        ShowError("Error loading image");
                        return;
                    }

                    // Close instruction
                    var closeLabel = new Label
                    {
                        Text = "Press ESC or click to close",
                        Font = new Font("Arial", 16),
                        ForeColor = Color.White,
                        BackColor = Color.Transparent,
                        AutoSize = true,
                        Padding = new Padding(10)
                    };

                    // Position close label at bottom center
                    closeLabel.Left = (form.Width - closeLabel.Width) / 2;
                    closeLabel.Top = form.Height - closeLabel.Height - 20;

                    form.Controls.Add(pictureBox);
                    form.Controls.Add(closeLabel);
                    closeLabel.BringToFront();

                    // Event handlers
                    form.KeyDown += (s, e) =>
                    {
                        if (e.KeyCode == Keys.Escape)
                            Hide();
                    };

                    form.Click += (s, e) => Hide();
                    pictureBox.Click += (s, e) => Hide();

                    lock (_lock)
                    {
                        _currentForm = form;
                    }

                    form.Show();
                    Application.Run(form);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error showing picture: {ex.Message}");
                }
            });
            thread.SetApartmentState(ApartmentState.STA);
            thread.IsBackground = true;
            thread.Start();
        }

        /// <summary>
        /// Shows a picture from byte array
        /// </summary>
        public void ShowPictureFromBytes(byte[] imageData, bool maintainAspectRatio = true, Color? backgroundColor = null)
        {
            try
            {
                // Save to temp file
                var tempPath = Path.Combine(Path.GetTempPath(), $"wrt_temp_{Guid.NewGuid()}.jpg");
                File.WriteAllBytes(tempPath, imageData);
                ShowPicture(tempPath, maintainAspectRatio, backgroundColor);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error showing picture from bytes: {ex.Message}");
            }
        }
    }
}
